engines/alcachofa/graphics-opengl-shaders.cpp

```diff
@@ -0,0 +1,255 @@
+		getQuadTexCoords(texMin, texMax, texCoords);
+		_vertices.push_back({ positions[0], texCoords[0], color });
+		_vertices.push_back({ positions[1], texCoords[1], color });
+		_vertices.push_back({ positions[2], texCoords[2], color });
+		_vertices.push_back({ positions[0], texCoords[0], color });
+		_vertices.push_back({ positions[2], texCoords[2], color });
+		_vertices.push_back({ positions[3], texCoords[3], color });
+	}
+
+	void setMatrices(bool flipped) override {
+		// adapted from https://en.wikipedia.org/wiki/Orthographic_projection
+		const float left = 0.0f;
+		const float right = _resolution.x;
+		const float bottom = flipped ? _resolution.y : 0.0f;
+		const float top = flipped ? 0.0f : _resolution.y;
+		const float near = -1.0f;
+		const float far = 1.0f;
+
+		_projection.setToIdentity();
+		_projection(0, 0) = 2.0f / (right - left);
+		_projection(1, 1) = 2.0f / (top - bottom);
+		_projection(2, 2) = -2.0f / (far - near);
+		_projection(3, 0) = -(right + left) / (right - left);
+		_projection(3, 1) = -(top + bottom) / (top - bottom);
+		_projection(3, 2) = -(far + near) / (far - near);
+	}
+
+private:
+	void checkFirstDrawCommand() {
+		OpenGLRenderer::checkFirstDrawCommand();
+
+		// submit batch
+		if (!_vertices.empty()) {
+			auto &vbo = _vbos[_curVBO];
+			_curVBO = (_curVBO + 1) % _vbos.size();
+
+			_shader.enableVertexAttribute("in_pos", vbo._bufferId, 2, GL_FLOAT, false, sizeof(Vertex), offsetof(Vertex, pos));
+			_shader.enableVertexAttribute("in_uv", vbo._bufferId, 2, GL_FLOAT, false, sizeof(Vertex), offsetof(Vertex, uv));
+			_shader.enableVertexAttribute("in_color", vbo._bufferId, 4, GL_UNSIGNED_BYTE, true, sizeof(Vertex), offsetof(Vertex, color));
+			_shader.use(true);
+
+			GL_CALL(glBindTexture(GL_TEXTURE_2D, _batchTexture == nullptr
+				? _whiteTexture->handle()
+				: _batchTexture->handle()));
+			GL_CALL(glBindBuffer(GL_ARRAY_BUFFER, vbo._bufferId));
+			if (vbo._capacity < _vertices.size()) {
+				vbo._capacity = _vertices.size();
+				glBufferData(GL_ARRAY_BUFFER, sizeof(Vertex) * _vertices.size(), _vertices.data(), GL_STREAM_DRAW);
+			} else
+				glBufferSubData(GL_ARRAY_BUFFER, 0, sizeof(Vertex) * _vertices.size(), _vertices.data());
+			glDrawArrays(GL_TRIANGLES, 0, _vertices.size());
+			_vertices.clear();
+		}
+
+		// setup next batch
+		setBlendFunc(_currentBlendMode);
+		_shader.setUniform("projection", _projection);
+		_shader.setUniform("blendMode", _currentTexture == nullptr ? 5 : (int)_currentBlendMode);
+		_shader.setUniform1f("lodBias", _currentLodBias);
+		_shader.setUniform("texture", 0);
+		_batchTexture = _currentTexture;
+	}
+
+	Matrix4 _projection;
+	Shader _shader;
+	Array<VBO> _vbos;
+	Array<Vertex> _vertices;
+	uint _curVBO = 0;
+	uint _usedTextureUnits = 0;
+	bool _needsNewBatch = false;
+	OpenGLTexture *_batchTexture = nullptr;
+	ScopedPtr<OpenGLTexture> _whiteTexture;
+
+	static constexpr const char *const kAttributes[] = {
+		"in_pos",
+		"in_uv",
+		"in_color",
+		nullptr
+	};
```

Comment:
This array is odr-used (its address is taken when it changes to const char* const* for loadFromStrings(...)), so the compiler emits a reference to a definition that the linker can’t find.

Either:
- Define it outside the class, like so:
```constexpr const char *const OpenGLRendererShaders::kAttributes[];```
OR:
- Add it locally inside the class initializer:
```OpenGLRendererShaders::OpenGLRendererShaders(Point resolution)
    : OpenGLRenderer(resolution) {
    static constexpr const char *const kAttributesLocal[] = {
        "in_pos", "in_uv", "in_color", nullptr
    };
    if (!_shader.loadFromStrings("alcachofa", kVertexShader, kFragmentShader, kAttributesLocal))
        error("Could not load shader");
}```

Reply:
done